torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1209,13 +1200,8 @@ bool ProcessGroupNCCL::dumpDebuggingInfo() {
     // their customized writer by inheriting `DebugInfoWriter` via
     // `registerDebugInfoWriter`.
     auto ncclTrace = dump_nccl_trace();
-    if (debugInfoWriter_ == nullptr) {
-      // Dump the trace blob into local disk as a fallback.
-      std::unique_ptr<DebugInfoWriter> debugInfoWriterPtr =
-          std::make_unique<DebugInfoWriter>(globalRank());
-      registerDebugInfoWriter(std::move(debugInfoWriterPtr));
-    }
-    debugInfoWriter_->write(ncclTrace);
+    DebugInfoWriter& writer = DebugInfoWriter::getWriter(globalRank());
```

Comment:
we should document the API behavior in the py doc.

iiuc there are a couple of non-obvious rules that are in play
1) if you register once, you can't register again (e.g. to change writer)
2) if you don't register before something calls dumpDebuggingInfo, then you lost your chance to register (and the filesystem writer will be auto-registered).  This should be fine as you're expected to register at trainer startup, before any dumping would occur.
3) we don't offer 'unregister' because it doesn't seem likely anyone would have that use case.

I think these are ok (at least for now), just document them.

Reply:
Got it, sounds good to me.